### PR TITLE
Optimise wiki.sortTiddlers()

### PR DESCRIPTION
**Pull request by Jermolene**
As discussed in #6042, there are some simple opportunities to make tiddler sorting more efficient.

These changes improve the startup time of a test wiki as follows:

* v5.1.23 mainRender: **5412.20ms**
* [575c233](https://github.com/Jermolene/TiddlyWiki5/commit/575c2335975cc6cc960d2061ba8fdfa726ed8618) mainRender: **11221.60ms**
* with this PR mainRender: **3862.00ms**
 
([575c233](https://github.com/Jermolene/TiddlyWiki5/commit/575c2335975cc6cc960d2061ba8fdfa726ed8618) is the last commit before reverting the problematic commit [dbd3f83](https://github.com/Jermolene/TiddlyWiki5/commit/dbd3f835bf8399ed1a3da7cc322ec9b6ab783d53))

### CODE REVIEW

**Comment by pmario** [comment]
There seems to be something wrong. The "sameday" filter should be only active 2 times. Now it is 4 

grafik

**Comment by Jermolene** [comment]
> There seems to be something wrong. The "sameday" filter should be only active 2 times. Now it is 4
> I'm using the test wiki from Hubert and this PR. Page reload with Recent tab selected

How would I replicate that test?

**Comment by Jermolene** [comment]
These changes are quite hard to review, but I'd really appreciate another pair of eyes if anyone else can have a look.

**Comment by saqimtiaz** [comment]
It is pretty tricky to review this as a diff/refactoring, so I looked through the new code on its own and it looks sound to me.

**Comment by Jermolene** [comment]
Thanks @saqimtiaz it is devilishly difficult; the existing logic has some unexpected fallthroughs...

I'd like to track down @pmario's problem before merging this, and then asking users for some wider testing.

**Comment by saqimtiaz** [comment]
> Thanks @saqimtiaz it is devilishly difficult; the existing logic has some unexpected fallthroughs...

If we don't already have tests for the edge cases then adding some would be one way to confirm that the new code behaves consistently with the old.

**Comment by Jermolene** [comment]
> If we don't already have tests for the edge cases then adding some would be one way to confirm that the new code behaves consistently with the old.

I have added some more tests as I traced through the code in 0be39cfbc2829dbd8637f35110b33c4bd2416dc3.

**Comment by pmario** [comment]
>How would I replicate that test?

 - I did download your `sort-optimisations` branch
 - created emtpy.html: `tiddlywiki .\editions\empty\ --build empty`
 - imported: `5.2.0-prerelease-perfm-5003_tids.html`
 - select: Recent tab
 - Save wiki
 - F12 dev console
 - F5 reload wiki
 - `$tw.perf.log()`

It's strange. To write this info I did do the steps and now I can't reproduce the problem anymore. 

But there is still the question, why is the filter executed twice instead of once directly after the startup.

grafik

File: core/modules/wiki.js [hunk 1]
```diff
@@ -365,49 +365,108 @@ Sort an array of tiddler titles by a specified field
...
+				if(!isCaseSensitive) {
+					a = a.toLowerCase();
+					b = b.toLowerCase();
+				}
+				return isDescending ? b.localeCompare(a) : a.localeCompare(b);
+			});
+		}
+	} else {
+		titles.sort(function(a,b) {
+			var x,y;
```
Comment (pmario) [inline core/modules/wiki.js:423]: if `a and b` would be initialized with `a="", b=""` here the else path in line 429 and 434 are redundant

**Comment by pmario** [comment]
In the `if(sortField === "title") {` path you did most of the isXX logic outside the sort function
in the `else {` path you do most of this logic inside the sort function again. ... Is there a reason for this difference?
I think the first approach is faster, since it contains less code in the repeated sort code

File: core/modules/wiki.js (hunk 1 above)
Comment (Jermolene) [reply core/modules/wiki.js:423]: Thanks that is true.

Comment (Jermolene) [reply core/modules/wiki.js:423]: Oh sorry, `a` and `b` are function parameters and so cannot be initialised like that.

**Comment by Jermolene** [comment]
> In the `if(sortField === "title") {` path you did most of the isXX logic outside the sort function
> in the `else {` path you do most of this logic inside the sort function again. ... Is there a reason for this difference?
> I think the first approach is faster, since it contains less code in the repeated sort code

I'm not sure that I understand. The final code block is the complete one, the preceding ones isolate special cases.

**Comment by Jermolene** [comment]
Thanks @pmario 

> But there is still the question, why is the filter executed twice instead of once directly after the startup.

Those per-filter timings are cumulative. If there were a subsequent refresh after the initial one then one would expect to see the number of executions of the filters to increase.

File: core/modules/wiki.js (hunk 1 above)
Comment (pmario) [reply core/modules/wiki.js:423]: uups you are right

**Comment by pmario** [comment]
>Those per-filter timings are cumulative. If there were a subsequent refresh after the initial one then one would expect to see the number of executions of the filters to increase.

That's right. But the only thing that happens is: Reload page ... So I would expect it to be executed only once, but is shows 2

**Comment by Jermolene** [comment]
> That's right. But the only thing that happens is: Reload page ... So I would expect it to be executed only once, but is shows 2

OK the question is whether you're seeing the same number of refreshes in the performance log for both of the test wikis.

Have you tried running the test rig? It does make this sort of thing a lot easier to manage.

**Comment by hubertgk** [comment]
Hi @Jermolene et al, when these changes are merged so as to be accessible on the [prerelease page](https://tiddlywiki.com/prerelease/) for download in an empty edition I'll be happy to test the version manually as before if this can help in any way (I don't know JS so I'm unable to contribute otherwise).

**Comment by pmario** [comment]
@hubertgk  ... I did upload a version with the new code for you to test here.
[empty.zip](https://github.com/Jermolene/TiddlyWiki5/files/7206003/empty.zip)

**Comment by Jermolene** [comment]
At the moment the blocker for merging is that I can't reproduce the problem reported by @pmario [here](https://github.com/Jermolene/TiddlyWiki5/pull/6053#issuecomment-922933299), if anyone else can help attempt to duplicate the problem I'd be very grateful.

**Comment by pmario** [comment]
>At the moment the blocker for merging is that I can't reproduce the problem reported by @pmario here, if anyone else can help attempt to duplicate the problem I'd be very grateful.

Sorry. This shouldn't be a blocker anymore. .. The problem I saw was the [4 times activation](https://github.com/Jermolene/TiddlyWiki5/pull/6053#issuecomment-922527140) of the filters. .. The 2 times are there since the performance log was activated. 

------

I'm only curious, why they are activated 2 times instead of once, if you do a page reload. .. But that shouldn't be a blocker

**Comment by saqimtiaz** [comment]
I can confirm there is no discrepancy in the number of times in the filters are run between 5.1.23, the current pre-release and this branch.

**Comment by Jermolene** [comment]
Thank you @pmario @saqimtiaz I'll merge this.

@hubertgk the changes will automatically appear on https://tiddlywiki.com/prerelease in 5-10 minutes.

**Comment by hubertgk** [comment]
Thank you @Jermolene and @pmario

I've tested my own wiki running the updated 5.2.0-prerelease on mobile just now and can confirm it's as snappy as it's ever been, perhaps even faster now than 5.1.23!

Thank you @Jermolene @saqimtiaz @pmario and everyone involved for your work on this issue.

**Comment by pmario** [comment]
>Thank you @Jermolene @saqimtiaz @pmario and everyone involved for your work on this issue.

Thank YOU for pointing out the issue ;)

**Comment by Jermolene** [comment]
That's great, thank you @hubertgk!
